invoker/buildpack.go

```diff
@@ -0,0 +1,148 @@
+	Error_UnsupportedLanguage = 105
+	Error_DetectInternalError = 106
+)
+
+type Buildpack interface {
+	Name() string
+	Detect(detect detect.Detect, metadata metadata.Metadata) (bool, error)
+	BuildPlan(detect detect.Detect, metadata metadata.Metadata) buildplan.BuildPlan
+	Invoker(build build.Build) (Invoker, bool, error)
+}
+
+type Invoker interface {
+	Contribute() error
+}
+
+type BuildpackCommands struct {
+	buildpack Buildpack
+}
+
+func NewBuildpackCommands(buildpack Buildpack) *BuildpackCommands {
+	return &BuildpackCommands{
+		buildpack: buildpack,
+	}
+}
+
+func (bc *BuildpackCommands) Name() string {
+	return bc.buildpack.Name()
+}
+
+func (bc *BuildpackCommands) Detect() {
+	detect, err := detect.DefaultDetect()
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to initialize Detect: %s\n", err)
+		os.Exit(Error_Initialize)
+	}
+
+	if err := detect.BuildPlan.Init(); err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to initialize Build Plan: %s\n", err)
+		os.Exit(Error_Initialize)
+	}
+
+	if code, err := bc.detect(detect); err != nil {
+		detect.Logger.Info(err.Error())
+		os.Exit(code)
+	} else {
+		os.Exit(code)
+	}
+}
+
+func (bc *BuildpackCommands) detect(detect detect.Detect) (int, error) {
+	metadata, ok, err := metadata.NewMetadata(detect.Application, detect.Logger)
+	if err != nil {
+		return detect.Error(Error_ReadMetadata), fmt.Errorf("unable to read riff metadata: %s", err.Error())
+	}
+
+	if !ok {
+		return detect.Fail(), nil
+	}
+
+	detected := false
+
+	if metadata.Override != "" {
+		if metadata.Override == bc.Name() {
+			detected = true
+			detect.Logger.Debug("Override language: %q.", bc.Name())
+		}
+	} else {
+		if detected, err = bc.buildpack.Detect(detect, metadata); err != nil {
+			detect.Logger.Info("Error trying to use %s invoker: %s", bc.Name(), err.Error())
+			return detect.Error(Error_DetectInternalError), nil
+		}
+
+		if detected {
+			detect.Logger.Debug("Detected language: %q.", bc.Name())
+		}
+	}
+
+	if detected {
+		return detect.Pass(bc.buildpack.BuildPlan(detect, metadata))
+	}
+
+	return detect.Fail(), nil
+}
+
+func (bc *BuildpackCommands) Build() {
+	build, err := build.DefaultBuild()
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to initialize Build: %s\n", err)
+		os.Exit(101)
+	}
+
+	if code, err := bc.build(build); err != nil {
+		build.Logger.Info(err.Error())
+		os.Exit(code)
+	} else {
+		os.Exit(code)
+	}
+}
+
+func (bc *BuildpackCommands) build(build build.Build) (int, error) {
+	build.Logger.FirstLine(build.Logger.PrettyIdentity(build.Buildpack))
+
+	if invoker, ok, err := bc.buildpack.Invoker(build); err != nil {
+		return build.Failure(105), err
+	} else if ok {
+		if err = invoker.Contribute(); err != nil {
+			return build.Failure(106), err
+		}
+		return build.Success(buildplan.BuildPlan{})
+	}
+
+	build.Logger.Info("Buildpack passed detection but did not know how to actually build. Should never happen.")
+	return build.Failure(104), nil
```

Comment:
```suggestion
	return build.Failure(Error_DetectInternalError), nil
```

Reply:
I created a var named `Error_BuildInternalError` since this is not part of the detect phase.